### PR TITLE
feat(file-upload): logic and validation changes

### PR DESCRIPTION
**Pull request by adisreyaj**
## ✨ Changes
image

- Updated the UI to be more clear about the file restrictions.
- Restriction messages are automatically build from the config file.

## ⚒️ Updates
- Introduced `SupportedFileType` enum with metadata about `extensions` and `mimeTypes`.
- Introduced  `UploaderConfig` interface for configuring the uploader.
	```ts
	export interface UploaderConfig {
	  supportedFileTypes: SupportedFileType[];
	  maxFileSizeInBytes: number;
	  maxNumberOfFiles: number;
	}
	```
- Restrictions are set on the input itself based on the config:
	- Limiting file selections
	- File Type Restrictions
- Validation checks made once files are selected/dropped and toast messages are shown for corresponding errors.
- Removed `ControlValueAccessor` implementation as the upload component requires a one-way communication only. ie it let the parent know about the file selected once the validations pass.
- Added test cases for all code paths.

### CODE REVIEW

File: projects/components/src/file-upload/file-types.ts [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+  [
+    SupportedFileType.Yaml,
+    {
+      extension: ['.yaml', '.yml'],
+      mimeTypes: ['application/x-yaml', 'text/yaml', 'text/x-yaml', 'application/yaml']
+    }
+  ]
+]);
+
+export namespace FileTypeUtil {
```
Comment (arjunlalb) [inline projects/components/src/file-upload/file-types.ts:23]: you'll have to suppress lint warning here

File: projects/components/src/file-upload/file-upload.component.scss [hunk 2]
```diff
@@ -57,3 +58,22 @@
...
+  padding: 0;
+  display: flex;
+  flex-direction: column;
+  gap: 4px;
+
+  .item {
+    display: flex;
+    align-items: center;
+    gap: 8px;
+    @include footnote($gray-4);
```
Comment (arjunlalb) [inline projects/components/src/file-upload/file-upload.component.scss:73]: inclusions first before overrides

File: projects/components/src/file-upload/file-upload.component.ts [hunk 3]
```diff
@@ -43,20 +38,46 @@ import { IconSize } from '../icon/icon-size';
...
-  public subText: string = '';
+  public subText?: string;
 
   @Input()
   public disabled: boolean = false;
 
+  @Input()
+  public config: UploaderConfig = {
+    maxNumberOfFiles: 10,
+    maxFileSizeInBytes: 20000000, // 20 MB
```
Comment (arjunlalb) [inline projects/components/src/file-upload/file-upload.component.ts:77]: nit: more readable if written as 20 * 1000 * 1000. also implicitly conveys the 20 MB part :)

File: projects/components/src/file-upload/file-upload.component.scss (hunk 2 above)
Comment (adisreyaj) [reply projects/components/src/file-upload/file-upload.component.scss]: Updated.

File: projects/components/src/file-upload/file-upload.component.ts (hunk 3 above)
Comment (adisreyaj) [reply projects/components/src/file-upload/file-upload.component.ts]: Updated.

File: projects/components/src/file-upload/file-type-pipe/file-type.pipe.ts [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+@Pipe({
+  name: 'htFilePipe'
+})
+export class FileTypePipe implements PipeTransform {
+  public transform(value: SupportedFileType | SupportedFileType[]): string {
+    if (Array.isArray(value)) {
+      return value.map(fileType => this.getFileExtension(fileType)).join(', ');
+    }
+
+    return this.getFileExtension(value);
```
Comment (itssharmasandeep) [inline projects/components/src/file-upload/file-type-pipe/file-type.pipe.ts:14]: Please add tests for this pipe

File: projects/components/src/file-upload/file-types.ts [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+  export const supportedFileExtensions = (fileTypes: SupportedFileType[]): string[] =>
+    fileTypes.flatMap(fileType => SUPPORTED_FILE_TYPE_METADATA_MAP.get(fileType)?.extension ?? []);
+
+  export const supportedFileExtensionsSet = (fileTypes: SupportedFileType[]): Set<string> =>
+    new Set(supportedFileExtensions(fileTypes));
+
+  export const supportedFileMimeTypes = (fileTypes: SupportedFileType[]): string[] =>
```
Comment (itssharmasandeep) [inline projects/components/src/file-upload/file-types.ts:31]: Let's add some tests for util as well

File: projects/components/src/file-upload/file-upload.component.ts [hunk 6]
```diff
@@ -126,13 +119,70 @@ export class FileUploadComponent implements ControlValueAccessor {
...
+    if (!this.validateFileTypes(fileList)) {
+      this.showFileTypeErrorToast();
+
+      return false;
+    }
+
+    return true;
+  }
+
+  public validateFileCount(fileList: FileList | null): boolean {
```
Comment (itssharmasandeep) [inline projects/components/src/file-upload/file-upload.component.ts:153]: isFileCountValid is better name for this, similarly for the functions below

Comment (itssharmasandeep) [reply projects/components/src/file-upload/file-upload.component.ts:153]: And these functions shouldn't be public

File: projects/components/src/file-upload/file-upload.component.ts [hunk 7]
```diff
@@ -43,20 +38,46 @@ import { IconSize } from '../icon/icon-size';
...
-  public subText: string = '';
+  public subText?: string;
 
   @Input()
   public disabled: boolean = false;
 
+  @Input()
+  public config: UploaderConfig = {
+    maxNumberOfFiles: 10,
+    maxFileSizeInBytes: 20 * 1000 * 1000, // 20 MB
```
Comment (itssharmasandeep) [inline projects/components/src/file-upload/file-upload.component.ts:77]:
```suggestion
    maxFileSizeInBytes: 20 * 1024 * 1024, // 20 MB
```

**Review by arjunlalb — APPROVED** [review]

**Review by arjunlalb — APPROVED** [review]

**Comment by anandtiwary** [comment]
oh, this one was stuck for a long time.

**Comment by itssharmasandeep** [comment]
> oh, this one was stuck for a long time.

It will be a breaking change for TR-UI

**Review by itssharmasandeep — APPROVED** [review]

**Pull request merged**